### PR TITLE
Refactor DPO data processing

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

This PR refactors DPO data processing to be less intricate and easier to understand and maintain. 

This main changes are the method `tokenize_row` and its equivalent for vision data `process_row`. 

This PR also solves a couple of issues:
- Wrong reported logits
- Wrong truncation logic

It also modifies the following:
- Make `max_length`, `max_prompt_length` and `max_completion_length` optional
- When merging prompt and completion alter tokenization

Regression tests are presented in my next message.

### Wrong truncation logic

Below is a simplified version of the current truncation logic (from `main`) used to handle sequences exceeding the allowed `max_length`:

```python
max_length = 10  # example value
max_prompt_length = 5  # example value

# if the combined sequence is too long, truncate the prompt
if len(prompt) + len(completion) > max_length:
    prompt = prompt[-max_prompt_length:]

# if it's still too long, truncate the completion from the end
if len(prompt) + len(completion) > max_length:
    completion = completion[: max_length - max_prompt_length]
```

#### Issues with this logic

1. **Unnecessary prompt truncation**: The prompt gets truncated more than necessary.
2. **Non-intuitive logic**: It's not immediately clear how truncation is applied.
3. **Lack of `max_completion_length` handling**: While `max_prompt_length` is applied, there's no equivalent handling for completions, making the logic even less intuitive.

#### Example

Consider the following:

```python
prompt = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
completion = [11, 12, 13, 14, 15]
```

Applying the original truncation logic gives:

```python
>>> prompt
[6, 7, 8, 9, 10]
>>> completion
[11, 12, 13, 14, 15]
>>> len(prompt) + len(completion)
10
```

In this case, the total length ends up being 10, even though `max_length` allows for up to 15 tokens. This means we could have preserved more of the prompt.

#### Proposed Solution

Here's a more flexible and intuitive approach (simplified for clarity):

```python
max_length = 10  # example value
max_prompt_length = None  # example value
max_completion_length = 8  # example value

# Truncate prompt if max_prompt_length is defined
if max_prompt_length is not None:
    prompt = prompt[-max_prompt_length:]

# Truncate completion if max_completion_length is defined
if max_completion_length is not None:
    completion = completion[:max_completion_length]

# Ensure combined prompt and completion don't exceed max_length
if len(prompt) + len(completion) > max_length:
    completion = completion[:max_length - len(prompt)]
```

#### Benefits

1. **Retains more of the prompt**: This logic avoids unnecessarily truncating the prompt.
2. **More intuitive**: It's clearer when each truncation step happens, and the flow is easier to follow.
3. **Supports flexible lengths**: Allows `None` for `max_length`, `max_prompt_length`, and `max_completion_length`, offering more control over how each part is truncated.

#### Result with the new logic

```python
>>> prompt, completion
([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], [11, 12, 13, 14, 15])
>>> len(prompt) + len(completion)
15
```

In this example, the prompt and completion are preserved within the allowed `max_length` without over-truncation.

### Wrong reported logits and log probs

The reported logits and log probs didn't take into account the padding. Eg:

```python
chosen_logits = logits.mean()
```

The new reporting first applies the mask before computing the mean:

```python
chosen_logits = logits[loss_mask].mean()
```

### When merging prompt and completion alter tokenization

Related #1960 and #2054

(very rare but worth mentioning)

There might be a difference between tokenizing the full concatenated text versus tokenizing the prompt and completion separately, and then merging their token lists.

Consider the following example:

```python
prompt = "Lorem."
completion = "\n\nipsum"
```

Tokenizing the combined prompt and completion produces different results than tokenizing each part separately and then concatenating their token lists:

```python
>>> tokenizer(prompt + completion)["input_ids"]
[32783, 382, 573, 1242]
>>> tokenizer(prompt)["input_ids"]
[32783, 13]
>>> tokenizer(completion)["input_ids"]
[271, 573, 1242]
```

This difference arises because we have a specific token for the string `".\n\n"`:

```python
>>> tokenizer.decode([382])
'.\n\n'
>>> tokenizer.decode([13, 271])
'.\n\n'
```

In the first tokenization method, the `".\n\n"` is treated as a single token, whereas in the second method, the token is split between the prompt and the completion.

In this example, the `main` code moves the merging token (`382`) to the completion:

```python
>>> prompt_ids = [32783]
>>> completion_ids = [382, 573, 1242]
>>> tokenizer.decode(prompt_ids)
'Lorem'
>>> tokenizer.decode(completion_ids)
'.\n\nipsum'
```

While this approach is functional, it slightly alters the original text since the final period is no longer part of the prompt. To avoid this, we can split the tokens without altering the data:

```python
>>> prompt_ids = [32783, 13]
>>> completion_ids = [271, 573, 1242]
>>> tokenizer.decode(prompt_ids)
'Lorem.'
>>> tokenizer.decode(completion_ids)
'\n\nipsum'
```

This way, the period remains in the prompt, preserving the original structure of the text. Note that this scenario is quite rare, it typically does not occur in conversational datasets.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qgallouedec** [comment]
Regression tests for:

- [ ] Ecoder-decoder

**Review by kashif — APPROVED** [review]

File: tests/test_dpo_trainer.py [hunk 1]
```diff
@@ -1139,7 +979,7 @@ def test_vdpo_trainer(self, model_id):
                 output_dir=tmp_dir,
                 per_device_train_batch_size=2,
                 max_length=512,
-                max_prompt_length=128,
```
Comment (qgallouedec) [inline tests/test_dpo_trainer.py:1142]: Otherwise it truncates the image tokens

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -815,27 +607,28 @@ def make_inputs_require_grad(module, input, output):
...
-                "model": model if self.is_encoder_decoder else None,
+                "processing_class": processing_class,
+                "max_prompt_length": args.max_prompt_length,
+                "max_completion_length": args.max_completion_length,
+                # for enc-dec, we add the special tokens ([bos_token] + prompt + [eos_token]; completion + [eos_token])
+                "add_special_tokens": self.is_encoder_decoder,
             }
             train_dataset = train_dataset.map(
-                _tokenize,
+                self.tokenize_row if not self.is_vision_model else self.process_row,
```
Comment (lewtun) [inline trl/trainer/dpo_trainer.py:617]: very clean separation of concerns!

File: trl/trainer/dpo_trainer.py [hunk 3]
```diff
@@ -815,27 +607,28 @@ def make_inputs_require_grad(module, input, output):
...
                 fn_kwargs=fn_kwargs,
-                batched=True,
                 num_proc=self.dataset_num_proc,
                 writer_batch_size=10,
                 desc="Tokenizing train dataset",
+                load_from_cache_file=False,
```
Comment (lewtun) [inline trl/trainer/dpo_trainer.py:622]: Is this to prevent the issue we've discussed previously about how changing a data mixture / number of samples doesn't recompute the processing?

File: trl/trainer/dpo_trainer.py [hunk 4]
```diff
@@ -815,27 +607,28 @@ def make_inputs_require_grad(module, input, output):
...
+                "max_prompt_length": args.max_prompt_length,
+                "max_completion_length": args.max_completion_length,
+                # for enc-dec, we add the special tokens ([bos_token] + prompt + [eos_token]; completion + [eos_token])
+                "add_special_tokens": self.is_encoder_decoder,
             }
             train_dataset = train_dataset.map(
-                _tokenize,
+                self.tokenize_row if not self.is_vision_model else self.process_row,
                 fn_kwargs=fn_kwargs,
-                batched=True,
```
Comment (lewtun) [inline trl/trainer/dpo_trainer.py:826]: Does this have any significant impact on preprocessing? I recall @kashif looked into this because the preprocessing was really slow for large preference datasets like Anthropic's HHH

File: trl/trainer/dpo_trainer.py [hunk 5]
```diff
@@ -1052,121 +933,116 @@ def null_ref_context(self):
             if self.ref_adapter_name:
                 self.model.set_adapter(self.model_adapter_name or "default")
 
-    def compute_reference_log_probs(self, padded_batch: Dict) -> Dict:
+    def compute_reference_log_probs(self, batch: Dict[str, torch.LongTensor]) -> Dict:
```
Comment (lewtun) [inline trl/trainer/dpo_trainer.py:936]: turbo nit: since we're already renaming variables, maybe we can align the method too?
```suggestion
    def compute_ref_log_probs(self, batch: Dict[str, torch.LongTensor]) -> Dict:
```

File: trl/trainer/dpo_trainer.py [hunk 6]
```diff
@@ -1284,190 +1162,148 @@ def dpo_loss(
...
+            attention_mask = torch.cat((prompt_attention_mask, completion_attention_mask), dim=1)
+            # Mask the prompt but not the completion for the loss
+            loss_mask = torch.cat(
+                (torch.zeros_like(prompt_attention_mask), completion_attention_mask),
+                dim=1,
+            )
 
-        if self.loss_type == "ipo":
-            all_logps = all_logps / size_completion
+            # Flush left to reduce the memory usage
```
Comment (lewtun) [inline trl/trainer/dpo_trainer.py:1232]: nice optimisation!

File: trl/trainer/dpo_trainer.py [hunk 7]
```diff
@@ -1284,190 +1162,148 @@ def dpo_loss(
...
+
+        output = {}
 
-        policy_weights = None
         if self.use_weighting:
-            chosen_weights = all_weights[:len_chosen]
-            rejected_weights = all_weights[len_chosen:]
-            policy_weights = torch.clamp(torch.exp(chosen_weights + rejected_weights), max=1)
+            with torch.no_grad():
+                # Eq (2) of the WPO paper: https://huggingface.co/papers/2406.11827
```
Comment (lewtun) [inline trl/trainer/dpo_trainer.py:1276]: Is this weighting applied for vanilla DPO as well or is it excluded since the loss doesn't depend on the `policy_weights` attribute?

**Review by lewtun — COMMENTED** [review]
Excellent refactor @qgallouedec - I really like how clean the new code is to read! I've left a few nits and questions about the weights loss from WPO, but otherwise this looks great. 

Regarding the difference in the chosen/rejected rewards of your regression tests, have you looked at the impact on downstream evals like IFEval / AlpacaEval / MixEval? I can run those for you if you have the checkpoints handy and then we can be pretty sure it's fine

File: trl/trainer/dpo_trainer.py [hunk 8]
```diff
@@ -76,266 +77,73 @@
...
+    Args:
+        pad_token_id (`int`):
+            Token ID to use for padding.
+        return_tensors (`str`, *optional*, defaults to `"pt"`):
+            Type of Tensor to return. Only `"pt"` is currently supported.
+
+    Examples:
+    ```python
+    >>> from trl import PreferenceCollator
+    >>> collator = PreferenceCollator(pad_token_id=0, collation_strategy="split")
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:95]: This is an option I deleted but forgot to remove from the doc, thanks!
```suggestion
    >>> collator = PreferenceCollator(pad_token_id=0)
```

File: trl/trainer/dpo_trainer.py (hunk 3 above)
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:622]: Arf no, it was more convenient to have it when debugging, good catch
```suggestion
```

File: trl/trainer/dpo_trainer.py [hunk 9]
```diff
@@ -815,27 +607,27 @@ def make_inputs_require_grad(module, input, output):
...
             if eval_dataset is not None:
                 eval_dataset = eval_dataset.map(
-                    _tokenize,
+                    self.tokenize_row if not self.is_vision_model else self.process_row,
                     fn_kwargs=fn_kwargs,
-                    batched=True,
                     num_proc=self.dataset_num_proc,
                     writer_batch_size=10,
                     desc="Tokenizing eval dataset",
+                    load_from_cache_file=False,
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:630]:
```suggestion
```

File: trl/trainer/dpo_trainer.py (hunk 5 above)
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py]: done in e509549

File: trl/trainer/dpo_trainer.py [hunk 10]
```diff
@@ -1478,67 +1314,42 @@ def get_batch_loss_metrics(
...
         reward_accuracies = (chosen_rewards > rejected_rewards).float()
 
         if self.args.rpo_alpha is not None:
-            # RPO loss from V3 of the paper:
-            losses = losses + policy_nll_loss * self.args.rpo_alpha
+            losses = losses + self.args.rpo_alpha * model_output["nll_loss"]  # RPO loss from V3 of the paper
 
         if self.use_weighting:
-            losses = losses * policy_weights
+            losses = losses * model_output["policy_weights"]
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:1333]: Answering https://github.com/huggingface/trl/pull/2209/files/9824f1670bb08845ab7bb850b5b14671bf843da5#r1804420747

I'm not 100% sure to understand your question. When the weighting is disabled, the loss won't be multiplied by the weights here. It's not even in `model_output`.
Please tell me if I've missed something.

File: trl/trainer/dpo_trainer.py (hunk 4 above)
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:826]: In fact, with batch processing, the problem is just shifted, as this function takes a list of examples instead of a single example, but processes it with a for loop.
The advantage of using `batched=False` is that you can make more intensive use of multiprocessing, which should be able to solve the slowness problem.

But I didn't know it was for that reason that we moved to `batched=True` (it wasn't mentioned in #1914). I'll double-check anyway to verify what I said.

Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:826]: |          | `dataset_num_proc = None` | `dataset_num_proc = 32` |
| -------- | ------------------------- | ----------------------- |
| `main`   | 04:59                     | 00:38                   |
| PR #2209 | 02:02                     | 00:17                   |

It's worth noting that the comparison isn't entirely fair, because in the case of the 2209, we don't compute the attention mask and the labels in the preprocessing function.

**Comment by qgallouedec** [comment]
> Regarding the difference in the chosen/rejected rewards of your regression tests, have you looked at the impact on downstream evals like IFEval / AlpacaEval / MixEval? I can run those for you if you have the checkpoints handy and then we can be pretty sure it's fine

Nice idea, I'll send you the checkpoints!

File: trl/trainer/dpo_trainer.py (hunk 4 above)
Comment (lewtun) [reply trl/trainer/dpo_trainer.py:826]: this looks great. let's avoid batching and just use multiprocessing as you have

File: trl/trainer/dpo_trainer.py (hunk 7 above)
Comment (lewtun) [reply trl/trainer/dpo_trainer.py:1274]: arf sorry i was being dumb - i missed the `self.use_weighting` line above and thought we were computing `policy_weights` by default. ignore my question :)

**Comment by qgallouedec** [comment]
@lewtun

Here is one: https://huggingface.co/qgallouedec/Qwen2.5-7B-DPO-2209
Trained with 

```
accelerate launch --config_file=examples/accelerate_configs/deepspeed_zero2.yaml examples/scripts/dpo.py \
    --dataset_name trl-lib/ultrafeedback_binarized \
    --model_name_or_path Qwen/Qwen2.5-7B-Instruct \
    --learning_rate 5.0e-7 \
    --num_train_epochs 1 \
    --gradient_checkpointing \
    --logging_steps 10 \
    --eval_strategy steps \
    --eval_steps 100 \
    --output_dir Qwen2.5-7B-DPO-2209 \
    --gradient_checkpointing \
    --max_length 512 \
    --use_peft \
    --bf16 \
    --push_to_hub
```

**Comment by qgallouedec** [comment]
## IFEval

The new implementation seems to improve results

| Model                                                                         | inst_level_loose_acc | inst_level_strict_acc | prompt_level_loose_acc | prompt_level_strict_acc |
| ----------------------------------------------------------------------------- | -------------------- | --------------------- | ---------------------- | ----------------------- |
| [Qwen2.5-7B](https://huggingface.co/Qwen/Qwen2.5-7B-Instruct)                 | 0.7122               | 0.6631                | 0.6026 ± 0.0211        | 0.5416 ± 0.0214         |
| [Qwen2.5-7B-DPO-main](https://huggingface.co/qgallouedec/Qwen2.5-7B-DPO-main) | 0.7182               | 0.6751                | 0.6155 ± 0.0209        | 0.5693 ± 0.0213         |
| [Qwen2.5-7B-DPO-2209](https://huggingface.co/qgallouedec/Qwen2.5-7B-DPO-2209) | 0.7326               | 0.6775                | 0.6303 ± 0.0208        | 0.5656 ± 0.0213         |

**Review by kashif — APPROVED** [review]
